lib/src/rules/use_late_for_private_fields_and_variables.dart

```diff
@@ -0,0 +1,175 @@
+      currentUnit = node.declaredElement;
+      lateables.putIfAbsent(currentUnit, () => []);
+      nullableAccess.putIfAbsent(currentUnit, () => {});
+
+      super.visitCompilationUnit(node);
+
+      final units = node.declaredElement.library.units;
+      if (units.every(lateables.keys.contains)) {
+        _checkAccess(units);
+
+        // clean up
+        units.forEach((unit) {
+          lateables.remove(unit);
+          nullableAccess.remove(unit);
+        });
+      }
+    }
+  }
+
+  @override
+  void visitNode(AstNode node) {
+    var element = DartTypeUtilities.getCanonicalElementFromIdentifier(node);
+    if (element != null) {
+      var parent = node.parent;
+      if (parent is Expression) {
+        parent = (parent as Expression).unParenthesized;
+      }
+      if (node is SimpleIdentifier && node.inDeclarationContext()) {
+        // ok
+      } else if (parent is PostfixExpression &&
+          parent.operand == node &&
+          parent.operator.type == TokenType.BANG) {
+        // ok non-null access
+      } else if (parent is AssignmentExpression &&
+          parent.operator.type == TokenType.EQ &&
+          context.typeSystem.isNonNullable(parent.rightHandSide.staticType)) {
+        // ok non-null access
+      } else {
+        nullableAccess[currentUnit].add(element);
+      }
+    }
+    super.visitNode(node);
+  }
+
+  @override
+  void visitFieldDeclaration(FieldDeclaration node) {
+    for (var variable in node.fields.variables) {
+      final parent = node.parent;
+      if (Identifier.isPrivateName(variable.name.name) ||
+          _isPrivateNamedCompilationUnitMember(parent) ||
+          _isPrivateExtension(parent)) {
+        _visit(variable);
+      }
+    }
+  }
+
+  @override
+  void visitTopLevelVariableDeclaration(TopLevelVariableDeclaration node) {
+    for (var variable in node.variables.variables) {
+      if (Identifier.isPrivateName(variable.name.name)) {
+        _visit(variable);
+      }
+    }
+  }
+
+  void _visit(VariableDeclaration variable) {
+    if (variable.isLate) {
+      return;
+    }
+    if (variable.isSynthetic) {
+      return;
+    }
+    if (context.typeSystem.isNonNullable(variable.declaredElement.type)) {
+      return;
+    }
+    lateables[currentUnit].add(variable);
+  }
+
+  void _checkAccess(List<CompilationUnitElement> units) {
+    final allNullableAccess =
+        units.expand((unit) => nullableAccess[unit]).toSet();
+    for (final unit in units) {
+      for (var variable in lateables[unit]) {
+        if (!allNullableAccess.contains(variable.declaredElement)) {
+          rule.reportLint(variable);
+        }
+      }
+    }
+  }
+}
+
+// see https://github.com/dart-lang/linter/pull/2189#issuecomment-660115569
+// Ideally we need to also ensure that there are no instances of either the
+// enclosing class or any subclass of the enclosing class that are ever
+// accessible outside this library.
+bool _isPrivateNamedCompilationUnitMember(AstNode parent) =>
```

Comment:
I think I didn't communicate clearly. Sorry about that.

What I was trying to say is that Erik pointed out a source of false positives that I suspect will result in enough false positives that we need to get rid of them. I think there are two ways to remove the false positives. One is by doing the work to prove that no instance of the class can escape the library, but that this approach is probably not worth the effort. The other is by not attempting to lint publicly named members of private classes.

The suggestion to add a comment was intended to capture the fact that _if_ we did the extra analysis we could extend the lint to also look for publicly named members of private classes.

Of course, I might be wrong about the false positives, but I don't know whether we have enough ported code to really gather data in this case.